### PR TITLE
Ana/162 bring back governance proposals

### LINKED ISSUE #162: bring back governance proposals
- list of governance proposals
- individual proposal pages

### PR DESCRIPTION
**Pull request by Bitcoinera**
Closes #162 

Bringing back governance :muscle:

### CODE REVIEW

**Comment by Bitcoinera** [comment]
Validators not loading now for some reason 

image

**Comment by jbibla** [comment]
you need to pull master in because the source files changed

**Comment by Bitcoinera** [comment]
> you need to pull master in because the source files changed

I already pulled. Maybe something went wrong in the merge process. There were a couple of conflicts

**Comment by Bitcoinera** [comment]
The only issue I see is that is takes a while to load proposals. I am going to change here the logic to always preload them non-blockingly when launching the app

**Comment by Bitcoinera** [comment]
OK, good to go! :dancers: 

So it currently looks like:

image


I will open a new PR with PageProposal complete

File: .eslintrc.js [hunk 1]
```diff
@@ -18,5 +18,6 @@ module.exports = {
   // add your custom rules here
   rules: {
     'no-console': 0,
+    'require-await': 0,
```
Comment (faboweb) [inline .eslintrc.js:21]: Remove

File: apis/cosmos-reducers.js [hunk 2]
```diff
@@ -691,7 +691,10 @@ export function validatorReducer(
     commission: Number(validator.commission.commission_rates.rate).toFixed(6),
     maxCommission: validator.commission.commission_rates.max_rate,
     maxChangeCommission: validator.commission.commission_rates.max_change_rate,
-    status: statusInfo.status,
+    status: {
+      value: statusInfo.status,
+      active: statusInfo.status === `ACTIVE`,
```
Comment (faboweb) [inline apis/cosmos-reducers.js:696]: Why this dupe?

File: apis/cosmos-source.js [hunk 3]
```diff
@@ -252,7 +254,7 @@ export default class CosmosAPI {
         BigNumber(tally.no).plus(tally.no_with_veto),
         totalVotingParticipation
       ),
-      links,
+      links: this.network.links,
```
Comment (faboweb) [inline apis/cosmos-source.js:257]: Rename to governanceLinks?

File: apis/cosmos-source.js [hunk 4]
```diff
@@ -402,27 +404,35 @@ export default class CosmosAPI {
...
       this.query('/distribution/community_pool'),
       this.getTopVoters(),
     ])
-    const communityPool = communityPoolArray
-      .map(this.reducers.coinReducer)
-      .find(({ denom }) => denom === network.stakingDenom)
+    const communityPool = communityPoolArray.find(
+      ({ denom }) =>
+        denom ===
+        this.network.coinLookup.find(
```
Comment (faboweb) [inline apis/cosmos-source.js:414]: Use gerCoinLookup

File: apis/cosmos-source.js [hunk 5]
```diff
@@ -402,27 +404,35 @@ export default class CosmosAPI {
...
+          ({ viewDenom }) => viewDenom === this.network.stakingDenom
+        ).chainDenom
+    ).amount
     return {
-      totalStakedAssets: setDecimalLength(
-        reducers.getStakingCoinViewAmount(totalBondedTokens),
-        2
+      totalStakedAssets: fixDecimalsAndRoundUpBigNumbers(
```
Comment (faboweb) [inline apis/cosmos-source.js:419]: Revert

File: apis/cosmos-source.js [hunk 6]
```diff
@@ -402,27 +404,35 @@ export default class CosmosAPI {
...
+        totalBondedTokens,
+        2,
+        this.network,
+        this.network.stakingDenom
       ),
       totalVoters: undefined,
-      treasurySize: setDecimalLength(
-        reducers.getStakingCoinViewAmount(communityPool),
-        2
+      treasurySize: fixDecimalsAndRoundUpBigNumbers(
```
Comment (faboweb) [inline apis/cosmos-source.js:426]: Revert

File: common/numbers.js [hunk 7]
```diff
@@ -1,17 +1,22 @@
+const BigNumber = require('bignumber.js')
 const SMALLEST = 1e-6
 // const language = window.navigator.userLanguage || window.navigator.language
 const language = `en` // TODO get from request, window is not available in SSR
 
+// truncate decimals to not round when using Intl.NumberFormat
+function truncate(number, digits) {
```
Comment (faboweb) [inline common/numbers.js:7]: Revert

File: common/numbers.js [hunk 8]
```diff
@@ -115,6 +120,34 @@ function bigFigureOrPercent(number) {
...
 }
 
+function fixDecimalsAndRoundUp(number, decimalsNumber) {
+  return (
+    (number.toFixed(decimalsNumber) * 10 ** decimalsNumber) /
+    10 ** decimalsNumber
+  )
+}
+
+function fixDecimalsAndRoundUpBigNumbers(
```
Comment (faboweb) [inline common/numbers.js:130]: Revert

File: components/governance/PageProposals.vue [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+        </div>
+        <div v-if="governanceOverview.totalVoters">
+          <h4>Total Voters</h4>
+          <p>{{ governanceOverview.totalVoters | prettyInt }}</p>
+        </div>
+      </div>
+    </div>
+
+    <!-- v-ifing ModalPropose until we enable the "propose" action in Substrate -->
+    <!-- <ModalPropose
```
Comment (faboweb) [inline components/governance/PageProposals.vue:39]: Remove

File: components/governance/ParticipantList.vue [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+      <li
+        v-for="(participant, index) in showingParticipants"
+        :key="index"
+        class="participant"
+      >
+        <div class="first-column">
+          <span class="icon">
+            <img v-if="participant.picture" :src="participant.picture" />
+            <!-- TODO: investigate why this doesn't work -->
+            <img v-else :src="network.icon" />
```
Comment (faboweb) [inline components/governance/ParticipantList.vue:14]: Use Avatar

File: components/governance/ParticipantList.vue [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+          <span v-if="participant.name" class="name">{{
+            participant.name
+          }}</span>
+          <span v-else class="name">{{
+            participant.address | formatAddress
+          }}</span>
+        </div>
+        <template v-if="participant.votingPower">
+          <div v-if="network.network_type === `cosmos`">
+            {{ participant.votingPower | bigFigureOrPercent }}
```
Comment (faboweb) [inline components/governance/ParticipantList.vue:25]: Simplify bigFigureOrPercent?

File: components/governance/ParticipantList.vue [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+    },
+  },
+  data: () => ({
+    showing: 5,
+    maxReached: false,
+    network,
+  }),
+  computed: {
+    showingParticipants() {
+      return JSON.parse(JSON.stringify(this.participants))
```
Comment (faboweb) [inline components/governance/ParticipantList.vue:87]: Why the cloning here?

File: components/governance/ParticipantList.vue [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+    moreAvailable() {
+      return this.showingParticipants.length < this.participants.length
+    },
+  },
+  methods: {
+    loadMore() {
+      if (!this.maxReached) {
+        this.showing += 5
+
+        if (
```
Comment (faboweb) [inline components/governance/ParticipantList.vue:100]: Why do we have a Max logic?

File: components/governance/ProposalDescription.vue [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+        </pre
+          >
+          <p v-else>{{ proposal.description }}</p>
+        </div>
+      </section>
+      <aside
+        v-if="supportingLinks && supportingLinks.length > 0"
+        class="supporting-links"
+      >
+        <h4>Supporting Links</h4>
```
Comment (faboweb) [inline components/governance/ProposalDescription.vue:23]: Are there supporting links for proposals now?

File: components/governance/ProposalHeader.vue [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+      <div class="header-top-row">
+        <div>
+          <BackButton />
+        </div>
+
+        <div class="inner-status">
+          <Status :status="status" />
+        </div>
+
+        <div class="buttons">
```
Comment (faboweb) [inline components/governance/ProposalHeader.vue:13]: Refactor to copy button

File: pages/governance/index.vue [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+      "
+      class="loading-row"
+    >
+      Loading...
+    </div>
+
+    <Card v-else-if="!proposals.length">
+      <div slot="title">No proposals</div>
+      <!-- TODO: add some actionable button / link -->
+      <div slot="subtitle">Submit now your own proposal to this blockchain</div>
```
Comment (faboweb) [inline pages/governance/index.vue:15]: Remove

File: pages/governance/index.vue [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+<script>
+import { mapState } from 'vuex'
+
+export default {
+  name: `proposals-index`,
+  computed: {
+    ...mapState('data', [`proposals`, `governanceOverview`]),
+    ...mapState(['session']),
+    oldChainDataMessage() {
+      return `If you expected to see proposals here that are missing,
```
Comment (faboweb) [inline pages/governance/index.vue:43]: This is not a thing for proposals

File: components/common/AppMenu.vue [hunk 18]
```diff
@@ -22,15 +22,15 @@
         <i class="material-icons notranslate">chevron_right</i>
       </nuxt-link>
 
-      <!-- <nuxt-link
-        class="app-menu-item "
-        to="/proposals"
+      <nuxt-link
+        class="app-menu-item"
+        to="/governance"
```
Comment (jbibla) [inline components/common/AppMenu.vue:27]:
```suggestion
        to="/proposals"
```

File: components/governance/ProposalHeader.vue [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+    governanceStatusEnum,
+  }),
+  computed: {
+    ...mapGetters([`currentNetwork`]),
+    currentRoute() {
+      return location.href
+    },
+    showVoteButton() {
+      // when the proposal is a Treasury proposal we won't show the Vote button
+      // for all Polkadot proposals we display the Vote button except for treasuries
```
Comment (jbibla) [inline components/governance/ProposalHeader.vue:93]: polka-no

File: network.js [hunk 20]
```diff
@@ -30,5 +30,24 @@ export default {
       ],
     },
   },
+  icon: '~/assets/images/currencies/atom.png',
+  governanceLinks: [
```
Comment (jbibla) [inline network.js:34]: should we include this or just delete it?

File: pages/governance/_id.vue [hunk 21]
```diff
@@ -0,0 +1,146 @@
+<template>
```
Comment (jbibla) [inline pages/governance/_id.vue:1]: i think it should be `pages/proposals/_id.vue`?

Comment (faboweb) [reply pages/governance/_id.vue:1]: Agreed

File: network.js (hunk 20 above)
Comment (faboweb) [reply network.js:34]: I was also thinking delete

File: components/governance/ProposalHeader.vue (hunk 19 above)
Comment (faboweb) [reply components/governance/ProposalHeader.vue:93]: Missed thar. Good eye

Comment (Bitcoinera) [reply components/governance/ProposalHeader.vue:93]: ouch. Removing

File: pages/governance/index.vue (hunk 17 above)
Comment (Bitcoinera) [reply pages/governance/index.vue]: But it happens the same for proposals, no? Or is that not the case?? :eyes: Are the cross-blockchain-version persistent?

File: components/common/AppMenu.vue (hunk 18 above)
Comment (Bitcoinera) [reply components/common/AppMenu.vue:27]: you need to rename the folder. I'm going to do it

File: components/governance/ProposalHeader.vue (hunk 19 above)
Comment (Bitcoinera) [reply components/governance/ProposalHeader.vue:93]: We should also actually remove the `isPolkadot` check  in `ImportSeedStep` and the `incorrectPolkadotSeed` from `FormMessage`. There are a couple of polka-no left around, as I see

Comment (Bitcoinera) [reply components/governance/ProposalHeader.vue:93]: Done

File: pages/governance/_id.vue (hunk 21 above)
Comment (Bitcoinera) [reply pages/governance/_id.vue:1]: Renaming

**Comment by Bitcoinera** [comment]
It seems there were some issues in the code review. We lost the right colors for `Rejected` and `Passed` and also the margins in `PageProposals`:

image

File: network.js (hunk 20 above)
Comment (Bitcoinera) [reply network.js]: Removed

**Comment by Bitcoinera** [comment]
Good to go! :woman_dancing:

**Comment by Bitcoinera** [comment]
oops, I did this the opposite that I wanted to... I wanted to actually merge this one into `162-finish-up-Pageproposal`.

OK, will close then that one and with this one should be enough

**Comment by Bitcoinera** [comment]
Although there seems to be some problem with ParticipantList in PageProposal (`_id`)

image

**Comment by faboweb** [comment]
> Although there seems to be some problem with ParticipantList in PageProposal (_id)

what is the issue?

**Comment by Bitcoinera** [comment]
> what is the issue?

OK, I think this is actually wanted. Validator images are not displaying but I think this is expected. You are using the Avatar component only, no `img`. Guess it is not so important?

**Pull request merged**